ntc_templates/templates/cisco_xr_show_bgp_vrf_all_neighbors_advertised-routes.textfsm

```diff
@@ -0,0 +1,20 @@
+Value NETWORK (\d+\.\d+\.\d+\.\d+)
+Value PREFIX_LENGTH (\d+)
+Value NEXT_HOP (\d+\.\d+\.\d+\.\d+)
+Value FROM ((\d+\.\d+\.\d+\.\d+)|\S+)
+Value AS_PATH (.*)
+Value Filldown ROUTE_DISTINGUISHER (\d+\:\d+)
+Value Filldown VRF (\S+)
+
+Start
+  ^\S{3}\s+\S{3}\s+\d{1,2}\s+\d+:\d+:\d+
+  ^Network\s+Next\s+Hop\s+From\s+AS\s+Path
+  ^Route\s+Distinguisher\:\s+${ROUTE_DISTINGUISHER}\s+\(default\s+for\s+vrf\s+${VRF}\)\s*$$ -> ROUTES
+  ^. -> Error
+
+ROUTES
+  ^${NETWORK}\/${PREFIX_LENGTH}\s+${NEXT_HOP}\s+${FROM}\s+${AS_PATH} -> Record
+  ^\s*$$ -> End
```

Comment:
I see what you did here, haha.
Hopscotch right to the `End` once you see a blank line. :grin:

Reply:
Way easier to finish the job :-)

Reply:
There is a disadvantage though in that once the first blank line is hit then it goes straight to ~jail~, err `End`. (Monopoly joke :laughing:)

I would say jumping to `End` in this manner may not be desirable though in the long term.

:grin: As long as it is ok with @jmcgill298 and the other maintainers then I suppose that's fine.

Reply:
ya, I'm not sure what this buys us, it seems there is just one more line to catch, so it should be easy enough to look for `^Processed`. We will also need to have an `^. -> Error` line in this section too.

Reply:
Given the syntax on lots of other templates, my $0.02 is not to redirect to end and instead handle the lines.

Allowing this state redirect to `End` does not align with the project requirement of matching all lines.

(Granted there are non-compliant templates from before that project requirement, but those will have to be retroactively identified. Holiday wish list [err, to-do list :laughing:] of sorts, haha.)

Reply:
Treating the last line is a pain... changed the template to address this. If this line does not goes to END, it creates an extra entry in the output, with all values blank.

If I proccess the line, correctly, treating prefixes and paths all anterior inputs will have these values blank, since it appears only on the last line.

Reply:
> Treating the last line is a pain... changed the template to address this. If this line does not goes to END, it creates an extra entry in the output, with all values blank.
> 
> If I proccess the line, correctly, treating prefixes and paths all anterior inputs will have these values blank, since it appears only on the last line.

@jorlandobr
I see, yes, there is an extra blank line.
If we can identify the start of a new VRF in that output that can be the trigger to change state back to `Start` state.

Might you have an XR example where there is more than one VRF?
(I would have to build something special for this. Though I may if you don't have something available.)

We'll want to make sure that this functions when there are one, two or more VRFs.
Blank line sending to `End` state may not work as we hope when there are multiple VRFs.

Thank you.

Reply:
Tomorrow I can provide an output of "vrf all". I'll try to find some box with few VRFs, or try to edit and sanitize the output.

I didn't procceded on that line ("vrf all") because the tests that my colleagues needed where VRF specific. A "vrf all" on our day to day basis is very polluted and almost useless.

Reply:
`show bgp vrf all neighbors 10.33.13.10 advertised-routes
Tue Nov 26 00:00:10.176 BRA

VRF: VRF:1250
-------------
% Neighbor not found

VRF: VMNET:1835
---------------
% Neighbor not found

VRF: VRFNAME:7848
----------------
% Neighbor not found

VRF: MONEY:5415
-----------------
% Neighbor not found

VRF: VRF2024:9005
-----------------
% Neighbor not found

VRF: COMPANY:154046
-----------------
% Neighbor not found

VRF: VRF_CIT:6910
-----------------
% Neighbor not found

VRF: VRFMAX:156010
------------------
% Neighbor not found

VRF: SCOTISH:1140
------------------
% Neighbor not found

VRF: VRF_NOVA:6114
------------------
% Neighbor not found

VRF: ABCD_UF:7990
------------------
% Neighbor not found

VRF: CITY_GER:7688
-------------------
% Neighbor not found

VRF: MARKET_INT_2:4776
---------------------
% Neighbor not found

VRF: NONAMENEWB:7609
---------------------
% Neighbor not found

VRF: VRF_CFTV:7129
----------------------
Network            Next Hop        From            AS Path
Route Distinguisher: 4230:7199 (default for vrf VRF_CFTV:7129)
0.0.0.0/0          172.23.33.9     10.43.78.102  4230 65000?
10.44.97.0/24      172.23.33.9     10.43.78.102  4230 65000?
10.44.255.184/29   172.23.33.9     10.43.78.102  4230 65000i
10.58.0.0/24       172.23.33.9     10.43.78.102  4230 64611i
10.58.8.0/22       172.23.33.9     10.43.78.102  4230 65000i
10.58.12.0/22      172.23.33.9     10.43.78.102  4230 65000i
10.58.16.0/22      172.23.33.9     10.43.78.102  4230 64612i
10.58.20.0/22      172.23.33.9     10.43.78.102  4230 65000i
10.58.24.0/22      172.23.33.9     10.43.78.102  4230 65000i
10.58.28.0/24      172.23.33.9     10.43.78.102  4230 65000i
10.58.32.0/22      172.23.33.9     10.43.78.102  4230 65000i
10.58.36.0/22      172.23.33.9     10.43.78.102  4230 65000i
10.58.40.0/22      172.23.33.9     10.43.78.102  4230 65000i
10.58.44.0/22      172.23.33.9     10.43.78.102  4230 64611i
10.58.48.0/22      172.23.33.9     10.43.78.102  4230 65000?
10.58.56.0/22      172.23.33.9     10.43.78.102  4230 65000?
10.58.64.0/22      172.23.33.9     10.43.78.102  4230 65000i
10.58.92.0/22      172.23.33.9     10.43.78.102  4230 65000?
10.58.127.0/29     172.23.33.9     10.43.78.102  4230 64611i
10.58.127.8/29     172.23.33.9     10.43.78.102  4230 65000i
10.58.127.16/28    172.23.33.9     10.43.78.102  4230 64612i
10.58.127.16/29    172.23.33.9     10.43.78.102  4230 65000i
10.58.127.32/29    172.23.33.9     10.43.78.102  4230 65000i
10.58.127.48/29    172.23.33.9     10.43.78.102  4230 65000i
10.58.127.56/29    172.23.33.9     10.43.78.102  4230 65000i
10.58.127.64/29    172.23.33.9     10.43.78.102  4230 65000i
10.58.127.72/29    172.23.33.9     10.43.78.102  4230 65000?
10.58.127.80/29    172.23.33.9     10.43.78.102  4230 65000i
10.58.127.88/29    172.23.33.9     10.43.78.102  4230 65000i
10.58.127.96/29    172.23.33.9     10.43.78.102  4230 65000i
10.58.127.120/29   172.23.33.9     10.43.78.102  4230 65000?
10.159.119.56/29   172.23.33.9     10.43.78.102  4230 65000?
10.159.220.0/23    172.23.33.9     10.43.78.102  4230 64611i
10.159.123.716/29  172.23.33.9     10.43.78.102  4230 65000i
10.244.34.0/29     172.23.33.9     10.43.78.102  4230 64611i
10.147.1.236/32   172.23.33.9     10.43.78.102  4230 65000 65022i
10.154.117.232/31 172.23.33.9     10.43.78.102  4230?
10.81.136.156/30  172.23.33.9     10.43.78.102  4230i
10.2.75.188/30    172.23.33.9     10.43.78.102  4230i
10.61.254.160/30  172.23.33.9     10.43.78.102  4230i
10.12.88.64/30    172.23.33.9     10.43.78.102  4230?
10.32..109.200/30  172.23.33.9     10.43.78.102  4230?
172.23.24.60/30    172.23.33.9     10.43.78.102  4230i
172.23.33.8/30     172.23.33.9     Local           4230?
10.68.161.24/30   172.23.33.9     10.43.78.102  4230?
10.68.161.28/30   172.23.33.9     10.43.78.102  4230?
10.127.111.208/30 172.23.33.9     10.43.78.102  4230i
192.168.34.36/30   172.23.33.9     10.43.78.102  4230?
192.168.27.244/30 172.23.33.9     10.43.78.102  4230?
192.168.65.80/30   172.23.33.9     10.43.78.102  4230i
192.168.16.0/30     172.23.33.9     10.43.78.102  4230?
192.168.43.140/30   172.23.33.9     10.43.78.102  4230?
192.168.80.212/30   172.23.33.9     10.43.78.102  4230i
192.168.110.40/30   172.23.33.9     10.43.78.102  4230?
192.168.79.184/30   172.23.33.9     10.43.78.102  4230?
192.168.147.204/30  172.23.33.9     10.43.78.102  4230i
192.168.215.12/30   172.23.33.9     10.43.78.102  4230?

Processed 57 prefixes, 57 paths

VRF: VRDUF_ACMEUF:9795
----------------------
% Neighbor not found

VRF: VRDUF_VRFCAR:9128
----------------------
% Neighbor not found

VRF: VOICE_CUST:5851
------------------------
% Neighbor not found

VRF: DATA_UFC_ROAD_12:6177
----------------------------
% Neighbor not found

VRF: DATA_UFC_NAME_XX:6741
----------------------------
% Neighbor not found`

Reply:
> Tomorrow I can provide an output of "vrf all". I'll try to find some box with few VRFs, or try to edit and sanitize the output.
> 
> I didn't procceded on that line ("vrf all") because the tests that my colleagues needed where VRF specific. A "vrf all" on our day to day basis is very polluted and almost useless.

This is helpful.
Yeah, I see what you're saying about lots of VRF that don't contain routes.

For those empty VRF, if it doesn't match the `Record` line they don't end up in the structured output. At least that's the way we have it currently and that's probably ideal. :grin:

Reply:
Our outputs are huge... usually 100's VRFs with 1000's routes each... I think that this one of the cases that we avoid "all" commands.

Without going to `End` still getting one empty register:
Using:
```
...
 ^Processed\s+
...
```

Last two registers:
```
	{
		"AS_PATH": "4230?",
		"FROM": "10.43.78.102",
		"NETWORK": "192.168.215.12",
		"NEXT_HOP": "172.23.33.9",
		"PREFIX_LENGTH": "30",
		"ROUTE_DISTINGUISHER": "4230:7199",
		"VRF": "VRF_CFTV:7129"
	},
	{
		"AS_PATH": "",
		"FROM": "",
		"NETWORK": "",
		"NEXT_HOP": "",
		"PREFIX_LENGTH": "",
		"ROUTE_DISTINGUISHER": "4230:7199",
		"VRF": "VRF_CFTV:7129"
	}
```


Using:

```
...
^Processed\s+ -> End
...
```

Last two registers:
```
	{
		"AS_PATH": "4230i",
		"FROM": "10.43.78.102",
		"NETWORK": "192.168.147.204",
		"NEXT_HOP": "172.23.33.9",
		"PREFIX_LENGTH": "30",
		"ROUTE_DISTINGUISHER": "4230:7199",
		"VRF": "VRF_CFTV:7129"
	},
	{
		"AS_PATH": "4230?",
		"FROM": "10.43.78.102",
		"NETWORK": "192.168.215.12",
		"NEXT_HOP": "172.23.33.9",
		"PREFIX_LENGTH": "30",
		"ROUTE_DISTINGUISHER": "4230:7199",
		"VRF": "VRF_CFTV:7129"
	}
```

Reply:
Looking into it.

Reply:
How do you generate the YML files from the JSON output? I am using nornir.tech to convert from JSON to YML. But after I have to check cases, spacing...

Reply:
@jorlandobr
There are development [helper scripts](https://github.com/networktocode/ntc-templates/blob/master/docs/dev/dev_parser.md#development-helper-scripts) you can use.

I have a fix for the blank entry that I'll be submitting shortly.

Reply:
> The helper scripts, I don't find it. To access them should I use git and download the project to my local machine?

@jorlandobr
Yeah, `git clone` your fork, switch to your branches where you're developing and run the helper script(s).

> I can't use git at work. Will try this at home.

You might be able to use [GitHub Codespaces](https://docs.github.com/en/codespaces/developing-in-a-codespace/creating-a-codespace-for-a-repository#creating-a-codespace-for-a-repository) from your browser, but there are ["core hours" limits per month](https://docs.github.com/en/billing/managing-billing-for-your-products/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#monthly-included-storage-and-core-hours-for-personal-accounts) based on plans. When you're not using your Codespace, you can stop it [here](https://github.com/codespaces) (or delete it when you're done with it).

I was able to get the below to function in case it is of use to you.

```zsh
# Go to your fork repo and choose the branch you're working on
# Click on the green Code button > Codespaces tab > Create codespace on <branch_name>
# Codespace launches in a new web browser tab
# From the VS Code Terminal
git switch <your_branch_name>
poetry shell
poetry install

# nothing but an example, hehe
python cli.py gen-yaml-folder -f tests/cisco_xr/show_ip_bgp_summary/
```

> The blank fix worked!

:100: :grin:
Yep, thank you, thank you, haha. :bow: